### PR TITLE
W7 - TienHuynh3 - Unit test

### PR DESCRIPTION
**Pull request by at-tienhuynh3**
#### Related issues (Optional)
Task #84 

#### Description (Required)
- [x] validation username
- [x] validation password
- [x] unit test

### CODE REVIEW

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/User.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package vn.asiantech.internship.unittest;
+
+/**
+ * Created by TienHuynh on 02/01/2018.
+ * AsianTech Co., Ltd
+ */
+public class User {
+    private String user;
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/User.java]: WARNING: [AvoidFieldNameMatchingTypeName] It is somewhat confusing to have a field name matching the declaring class name
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#AvoidFieldNameMatchingTypeName

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/test/PassWordUnitTest.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PassWordUnitTest {
+
+    @Mock
+    private User mUser;
+
+    @Test
+    public void checkDifferentUserName() {
+        Mockito.when(mUser.getUser()).thenReturn("tienhuynh");
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/test/PassWordUnitTest.java]: WARNING: [AvoidDuplicateLiterals] The String literal "tienhuynh" appears 5 times in this file; the first occurrence is on line 25
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/test/UserUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+public class UserUnitTest {
+
+    @Mock
+    private User mUser;
+
+    @Test
+    public void checkLengthUserName() {
+        Mockito.when(mUser.getUser()).thenReturn("tienhuuynh");
+        Assert.assertTrue(UserNameValidation.checkLengthUserName(mUser.getUser()));
+        Mockito.when(mUser.getUser()).thenReturn("tien");
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/test/UserUnitTest.java]: WARNING: [AvoidDuplicateLiterals] The String literal "tien" appears 4 times in this file; the first occurrence is on line 27
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

**Comment by at-tienhuynh3** [comment]
@at-trungnguyen pls review for me. Thank you <3

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/test/UserUnitTest.java (hunk 3 above)
Comment (at-tienhuynh3) [reply app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/test/UserUnitTest.java]: i fixed it

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/User.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package vn.asiantech.internship.unittest;
+
+/**
+ * Created by TienHuynh on 02/01/2018.
+ * AsianTech Co., Ltd
+ */
+public class User {
+
+    private String user;
```
Comment (at-dev-support-01) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/User.java]: WARNING: [AvoidFieldNameMatchingTypeName] It is somewhat confusing to have a field name matching the declaring class name
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#AvoidFieldNameMatchingTypeName

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            @Override
+            public void onClick(View view) {
+                // validate
+                validate();
+            }
+
+            /**
+             * this method used to validate EditText
+             */
+            private void validate() {
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java]: move this method to outside

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                    showToast(getString(R.string.validate_done));
+                }
+            }
+
+            /**
+             * Validate UserName
+             *
+             * @param userName
+             */
+            private boolean validateUserName(String userName) {
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java]: move this method to outside

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                if (validateUserName(mEdtUserName.getText().toString()) &&
+                        validatePassWord(mEdtPassWord.getText().toString())) {
+                    showToast(getString(R.string.validate_done));
+                }
+            }
+
+            /**
+             * Validate UserName
+             *
+             * @param userName
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java]: change to `@param userName username`is better avoid warning.

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+             */
+            private boolean validateUserName(String userName) {
+                boolean validationUserName = false;
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java]: `boolean` variable should start by `is` ->`isValidateUserName`

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+                    showToast(getString(R.string.validate_username_special_char));
+                } else if (!UserNameValidation.checkUserNameIgnoreUpperCase(userName)) {
+                    showToast(getString(R.string.validate_username_no_distinction));
+                } else {
+                    validationUserName = true;
+                }
+                return validationUserName;
+            }
+
+            /**
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java]:
```/**
             * Validate for password
             *
             * @param password password
             * @return is validate
             */
```

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+             * Validation Password
+             *
+             * @param password
+             * @return
+             */
+            private boolean validatePassWord(String password) {
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java]: move to outside

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+                boolean validationPassword = false;
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java]: start by `is`

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+                    showToast(getString(R.string.validate_password_repeat));
+                } else {
+                    validationPassword = true;
+                }
+                return validationPassword;
+            }
+        });
+    }
+
+    /**
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/LoginUnitTestActivity.java]: fix same above

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/PasswordValidation.java [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+import java.util.regex.Pattern;
+
+/**
+ * Created by TienHuynh on 02/01/2018.
+ * AsianTech Co., Ltd
+ */
+public final class PasswordValidation {
+
+    /**
+     * Check password different username
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/PasswordValidation.java]: change comment same above, avoid warning

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/PasswordValidation.java [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     * @param userName
+     * @param password
+     * @return
+     */
+    public static boolean checkDifferentUserName(String userName, String password) {
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/PasswordValidation.java]: method name should be `isDifferentUserName`

Comment (at-trungnguyen) [reply app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/PasswordValidation.java]: Fix for all method return `boolean`

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/UserNameValidation.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.Locale;
+
+/**
+ * Created by TienHuynh on 02/01/2018.
+ * AsianTech Co., Ltd
+ */
+public final class UserNameValidation {
+
+    /**
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/UserNameValidation.java]: fix this file same above

File: app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/UserNameValidation.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+        return !userName.contains(" ");
+    }
+
+    /**
+     * check user name have special char
+     *
+     * @param userName
+     * @return
+     */
+    public  static boolean checkUserNameSpecialChar(String userName) {
```
Comment (at-trungnguyen) [inline app/src/at-tienhuynh/java/vn/asiantech/internship/unittest/UserNameValidation.java]: format this file

**Comment by at-trungnguyen** [comment]
@at-tienhuynh3 You wrote `unit test` wrong position. Please create new`test` directory in `src` root.

File: app/src/testAt-tienhuynhDebug/java/PassWordUnitTest.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+import org.mockito.runners.MockitoJUnitRunner;
+
+import vn.asiantech.internship.unittest.PasswordValidation;
+import vn.asiantech.internship.unittest.User;
+
+/**
+ * Created by TienHuynh on 02/01/2018.
+ * AsianTech Co., Ltd
+ */
+@RunWith(MockitoJUnitRunner.class)
```
Comment (at-trungnguyen) [inline app/src/testAt-tienhuynhDebug/java/PassWordUnitTest.java]: wrong import, it's deprecated.

File: app/src/testAt-tienhuynhDebug/java/PassWordUnitTest.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+import org.mockito.runners.MockitoJUnitRunner;
+import vn.asiantech.internship.unittest.PasswordValidation;
+import vn.asiantech.internship.unittest.User;
+
+/**
+ * Created by TienHuynh on 02/01/2018.
+ * AsianTech Co., Ltd
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PassWordUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-tienhuynhDebug/java/PassWordUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

File: app/src/testAt-tienhuynhDebug/java/UserUnitTest.java [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+import org.mockito.junit.MockitoJUnitRunner;
+import vn.asiantech.internship.unittest.User;
+import vn.asiantech.internship.unittest.UserNameValidation;
+
+/**
+ * Created by TienHuynh on 02/01/2018.
+ * AsianTech Co., Ltd
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class UserUnitTest {
```
Comment (at-dev-support-01) [inline app/src/testAt-tienhuynhDebug/java/UserUnitTest.java]: WARNING: [NoPackage] All classes and interfaces must belong to a named package
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#NoPackage

**Pull request merged**